feat: Reconcile Payments in background

ruthra-kumar opened the pull request:
For extremely large volumes of Invoices and Payments, Payment Reconciliation tool may timeout. This feature allows for running the same reconciliation tool in the background. 

Documentation can be found [here](https://docs.erpnext.com/docs/v14/user/manual/en/accounts/semi-auto-payment-reconciliation)

Changes:
1. 2 new doctypes: `Process Payment Reconciliation` and `Process Payment Reconciliation Log` have been introduced. These reuse the existing reconciliation logic. `Process Payment Reconciliation` will have similar filters as Payment Reconciliation. Once submitted, 'Reconcile in Background' button will be availabe. Once triggered, `Process Payment Reconciliation Log` will be created. Payments are allocated in FIFO order. 
<img width="1440" alt="Screenshot 2023-04-04 at 5 34 07 PM" src="https://user-images.githubusercontent.com/3272205/229786964-3fe84fca-8bc7-4d6d-9cd8-fb3efa1941c1.png">
<img width="1440" alt="Screenshot 2023-04-04 at 5 34 28 PM" src="https://user-images.githubusercontent.com/3272205/229786987-307e6448-4584-4fbc-8f66-6959d7416837.png">
<img width="1440" alt="Screenshot 2023-04-04 at 5 34 33 PM" src="https://user-images.githubusercontent.com/3272205/229786995-eba392e2-afb0-43c1-be3a-41e4c7bcfc69.png">
<img width="1440" alt="Screenshot 2023-04-04 at 5 34 39 PM" src="https://user-images.githubusercontent.com/3272205/229786998-5ca1d410-0796-451d-85d7-69b52bb804a0.png">
<img width="1440" alt="Screenshot 2023-04-04 at 5 35 01 PM" src="https://user-images.githubusercontent.com/3272205/229787002-5290cf5b-dab2-4a56-9dfb-7bc20e4e2be5.png">

2. Added a checkbox in `Accounts Settings` to enable/disable this feature 
3. Added few validations in Payment Reconciliation tool to check for running jobs with same filter. 
<img width="1440" alt="Screenshot 2023-04-04 at 5 37 58 PM" src="https://user-images.githubusercontent.com/3272205/229787143-020fdd3a-dc6c-4790-881a-36b8636fc8b1.png">
<img width="1440" alt="Screenshot 2023-04-04 at 5 38 36 PM" src="https://user-images.githubusercontent.com/3272205/229787159-a95215f0-789a-462e-b0a7-34c5d9bc1c00.png">

ruthra-kumar committed:
feat: auto reconcile in background

ruthra-kumar committed:
test: create a bunch of vouchers for testing auto reconcile

ruthra-kumar committed:
refactor: check for running job while using reconciliation tool

ruthra-kumar committed:
chore: Option to enable auto reconciliation in settings

ruthra-kumar committed:
refactor: validate if feature is enabled in settings

deepeshgarg007 commented on erpnext/accounts/doctype/accounts_settings/accounts_settings.json:
```diff
@@ -360,14 +357,29 @@
+   "default": "0",
+   "fieldname": "enable_auto_reconciliation",
+   "fieldtype": "Check",
+   "label": "Enable Auto Reconciliation"
```
Explicitly specify payment reconciliation?

ruthra-kumar replied:
Yes. I mean, by default not every one needs this feature. Moreover, as this is still in development, better to let the user enable this explicitly if they want it.

deepeshgarg007 replied:
No, I meant the label, "Enable Auto Reconciliation" is ambiguous, it could be bank or payment

ruthra-kumar replied:
Label has been changed

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2023, Frappe Technologies Pvt. Ltd. and contributors
+// For license information, please see license.txt
+
+frappe.ui.form.on("Auto Reconcile", {
```
"Process Payment Reconciliation" keeping the currency naming convention like "Process Deferred Accounting" and "Process Loan Interest Accrual"?

ruthra-kumar replied:
Huh.. It does have a common theme to it. I'll rename both of them.

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+		frm.set_query('receivable_payable_account',  function(doc, cdt, cdn) {
+			return {
+				filters: {
+					"company": locals[cdt][cdn].company,
```
```suggestion
					"company": doc.company,
```

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+		frm.set_query('cost_center', function(doc, cdt, cdn) {
+			return {
+				filters: {
+					"company": locals[cdt][cdn].company,
```
```suggestion
					"company": doc.company,
```

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+				}
+			}
+		});
+		frm.set_query('receivable_payable_account',  function(doc, cdt, cdn) {
```
Such queries should be under `refresh`. They should refresh if the company is changed

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+		frm.set_query('bank_cash_account', function(doc, cdt, cdn) {
+			return {
+				filters:[
+					['Account', 'company', '=', locals[cdt][cdn].company],
```
```suggestion
					['Account', 'company', '=', doc.company],
```

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.js:
```diff
@@ -0,0 +1,64 @@
+	},
+	refresh: function(frm) {
+		if (frm.doc.docstatus==1 && ['Queued', 'Running'].find(x => x == frm.doc.status)) {
+			var execute_btn = __("Reconcile in Background")
```
```suggestion
			let execute_btn = __("Reconcile in Background")
```

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.json:
```diff
@@ -0,0 +1,172 @@
+ "sort_field": "modified",
+ "sort_order": "DESC",
+ "states": [],
+ "title_field": "company"
```
Should have a naming series, right? Why title with company?

ruthra-kumar replied:
Naming Series added

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+		if self.receivable_payable_account:
+			if not frappe.db.get_list(
+				"Account", filters={"name": self.receivable_payable_account, "company": self.company}
+			):
```
```suggestion
			if self.company != frappe.db.get_value("Account", receivable_payable_account, "company"):
```

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+
+	def validate_bank_cash_account(self):
+		if self.bank_cash_account:
+			if not frappe.db.get_list(
```
Update same as suggested above?

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+					)
+				)
+
+	def before_save(self):
+		self.status = "Draft"
+		self.error_log = ""
```
By default status for saved doc is "Draft", do we need this?

ruthra-kumar replied:
Removed it.

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+		self.status = "Draft"
+		self.error_log = ""
+
+	def before_submit(self):
```
Shouldn't this be `on_submit`?

ruthra-kumar replied:
Using `on_submit` controller now

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+def get_progress(docname: str | None = None) -> float:
+	progress = 0.0
+	if docname:
+		reconcile_log = frappe.db.get_list(
+			"Auto Reconcile Log", filters={"auto_reconcile": docname}, as_list=1
+		)
```
Use `get_value`?

ruthra-kumar replied:
Done

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+			"Auto Reconcile Log", filters={"auto_reconcile": docname}, as_list=1
+		)
+		if reconcile_log:
+			processed = frappe.db.get_value("Auto Reconcile Log", reconcile_log[0][0], "reconciled_entries")
+			total = frappe.db.get_value("Auto Reconcile Log", reconcile_log[0][0], "total_allocations")
```
Merge in a single query

ruthra-kumar replied:
merge?

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+
+
+def get_pr_instance(doc: str):
+	auto_reconcile = frappe.get_doc("Auto Reconcile", doc)
```
Use get_value instead of get_doc, not using any child table values so better to use get_value instead of get_doc

ruthra-kumar replied:
This doctype doesn't have any child table. It only has the filters.

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+def is_job_running(job_name: str) -> bool:
+	jobs = frappe.db.get_list("RQ Job", filters={"status": ["in", ["started", "queued"]]})
+	for x in jobs:
+		if x.job_name == job_name:
```
`get_list` by default will get name field I guess, or will it return `job_name`.
Also can be done by a single `get_value` query

ruthra-kumar replied:
Parameters like "fields" are not yet implemented for `RQ Job` virtual doctype. Normal `get_all` output includes all the fields.
`get_value` is unsupported as well.
<img width="500" alt="Screenshot 2023-04-01 at 2 55 14 PM" src="https://user-images.githubusercontent.com/3272205/229277918-f222b6c7-c312-4beb-8b46-ac747ca47570.png">

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile/auto_reconcile.py:
```diff
@@ -0,0 +1,407 @@
+		res = is_any_doc_running()
+		if not res:
+			# Get Queued item and start the process
+			queued = frappe.db.get_list(
```
`get_all` is preferred on server side, `get_list` does permission checks and can cause unwanted errors in some cases

ruthra-kumar replied:
replaced with `get_all` as suggested

ruthra-kumar replied:
addressed

deepeshgarg007 commented on erpnext/accounts/doctype/auto_reconcile_log/auto_reconcile_log.js:
```diff
@@ -0,0 +1,14 @@
+
+frappe.ui.form.on("Auto Reconcile Log", {
+	refresh(frm) {
+		var progress = 0;
```
```suggestion
		let progress = 0;
```

ruthra-kumar replied:
addressed

ruthra-kumar committed:
chore: using doc to get filter values

ruthra-kumar committed:
chore: use frappe.db.get_value in validations

ruthra-kumar committed:
chore: cleanup commented out code

ruthra-kumar committed:
chore: replace get_list with get_all

ruthra-kumar committed:
chore: use block scope variable

ruthra-kumar committed:
chore: type information for functions

ruthra-kumar committed:
refactor: flag to ignore job validation check

ruthra-kumar committed:
chore: create test_records file

ruthra-kumar committed:
refactor: update parent doc status if all reconciled

ruthra-kumar committed:
chore: renamed auto_reconcile to process_payment_reconciliation

ruthra-kumar committed:
chore: another child doctype to hold payments

ruthra-kumar committed:
chore: remove duplicate field

ruthra-kumar committed:
chore: add fetched payments to log

ruthra-kumar committed:
chore: Popup comment message update

ruthra-kumar committed:
chore: replace get_all with get_value

ruthra-kumar committed:
chore: replace label in settings page

ruthra-kumar committed:
chore: remove unit test and records

ruthra-kumar committed:
refactor: status in reconciliation log

ruthra-kumar committed:
refactor: set status in log as well

ruthra-kumar committed:
chore: fix field name

ruthra-kumar committed:
chore: change triggered job name

ruthra-kumar committed:
chore: use status field in list view of log

ruthra-kumar committed:
chore: status while there are no allocations

ruthra-kumar committed:
refactor: split trigger function into two

ruthra-kumar committed:
chore: adding cancelled status

ruthra-kumar committed:
refactor: function trigger queued docs

ruthra-kumar committed:
chore: cron job scheduled

ruthra-kumar committed:
chore: fixing accouts settings json file

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+		if self.receivable_payable_account:
+			if self.company != frappe.db.get_value("Account", self.receivable_payable_account, "company"):
+				frappe.throw(
+					_("Receivable/Payable Account: {0} doesn't below to company {1}").format(
```
```suggestion
					_("Receivable/Payable Account: {0} doesn't belong to company {1}").format(
```

ruthra-kumar replied:
typo fixed

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+		if self.bank_cash_account:
+			if self.company != frappe.db.get_value("Account", self.bank_cash_account, "company"):
+				frappe.throw(
+					_("Bank/Cash Account {0} doesn't below to company {1}").format(
```
```suggestion
					_("Bank/Cash Account {0} doesn't belong to company {1}").format(
```

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+
+
+@frappe.whitelist()
+def get_progress(docname: str | None = None) -> float:
```
We have `frappe.publish_realtime` to update progress, tried that instead of this approach?

ruthra-kumar replied:
`frappe.publish_realtime` would be unsuitable in this case. We need a listener and the publishing event should happen in realtime. Say if the job paritally failed, then there won't be any progress published. Or even if the Job ran successfully, then there won't be job to publish the progress.

The log `Process Payment Reconciliation Log` has `reconciled_entries` variable which holds count of successfully reconciled entries. Just querying that is enough to know the progress.

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation_list.js:
```diff
@@ -0,0 +1,14 @@
+frappe.listview_settings['Process Payment Reconciliation'] = {
+	add_fields: ["status"],
+	get_indicator: function(doc) {
+		var colors = {
```
```suggestion
		let colors = {
```

ruthra-kumar replied:
changed

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+	if for_filter:
+		if type(for_filter) == str:
+			for_filter = frappe.json.loads(for_filter)
+		running_doc = frappe.db.get_all(
```
Use `frappe.db.exists` or `frape.db.get_value` instead? `get_all` seems unnecessary

ruthra-kumar replied:
replaced with get_value

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+			as_list=1,
+		)
+	else:
+		running_doc = frappe.db.get_all(
```
Use `frappe.db.exists` or `frape.db.get_value` instead? `get_all` seems unnecessary

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.js:
```diff
@@ -0,0 +1,71 @@
+	},
+	refresh: function(frm) {
+		if (frm.doc.docstatus==1 && ['Queued', 'Running'].find(x => x == frm.doc.status)) {
+			var execute_btn = __("Reconcile in Background")
```
```suggestion
			let execute_btn = __("Reconcile in Background")
```

ruthra-kumar replied:
changed

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+
+	def on_submit(self):
+		self.db_set("status", "Queued")
+		self.db_set("error_log", None)
```
There is also a log document, right? we can log the errors there itself

ruthra-kumar replied:
Yes. But, if the log document large(large no of allocation) just opening the log itself will be time consuming. So, i think its better to update in Parent doc as well. So, if the job fails, the error msg can be easily identified here

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+			"Process Payment Reconciliation Log", filters={"process_pr": docname}, fieldname="name"
+		)
+		if reconcile_log:
+			processed = frappe.db.get_value(
+				"Process Payment Reconciliation Log", reconcile_log, "reconciled_entries"
+			)
+			total = frappe.db.get_value(
+				"Process Payment Reconciliation Log", reconcile_log, "total_allocations"
+			)
```
Can be fetched in a single query?

ruthra-kumar replied:
Addressed

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+		frappe.msgprint(_("Scheduler is Inactive. Can't trigger job now."))
+
+
+@frappe.whitelist()
+def trigger_reconciliation_for_queued_docs():
```
Only trigger for queued docs, needs whitelisting?

ruthra-kumar replied:
Removed

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+	if not frappe.db.get_single_value(
+		"Accounts Settings", "enable_payment_reconciliation_in_background"
+	):
+		frappe.throw(
+			_(
+				"Payment Reconciliation through backgound Job has been disabled. Enable it through {}"
+			).format(get_link_to_form("Accounts Settings", "Accounts Settings"))
+		)
```
Jobs will result in an error if settings is disabled, should simply return

ruthra-kumar replied:
If it just returns, the  Cron Job will show up as Completed. User will have no way to know that the feature was disabled. The `Process Payment Reconciliation` doc, which will be in queued status, will also not have any indication that Job didn't run due to disabled feature.

I think it's better to throw an exception here.

deepeshgarg007 commented on erpnext/accounts/doctype/process_payment_reconciliation/process_payment_reconciliation.py:
```diff
@@ -0,0 +1,500 @@
+					doc=doc,
+				)
+		else:
+			reconcile_log = frappe.get_doc("Process Payment Reconciliation Log", log[0][0])
```
You can simply pluck name, won't have to do log[0][0

ruthra-kumar replied:
Addressed

ruthra-kumar committed:
chore: typos and variable scope

ruthra-kumar committed:
chore: use 'pluck' in db call

ruthra-kumar committed:
chore: remove redundant whitelist decorator

ruthra-kumar committed:
chore: use single DB call to fetch values

ruthra-kumar committed:
chore: replace get_all with get_value

ruthra-kumar committed:
refactor: use raw db calls to fetch reconciliation log records

Using get_doc on `Process Payment Reconciliation Log` is costly when
handling large volumes of invoices.

Use raw frappe.db.get_all to selectively pull status and reconciled count

ruthra-kumar committed:
chore: update status on successful batch operation

ruthra-kumar committed:
chore: make payment table readonly

ruthra-kumar committed:
chore: ability to pause the background job

ruthra-kumar committed:
chore: remove isolate_each_allocation

ruthra-kumar committed:
chore: more description in progress bar

ruthra-kumar committed:
refactor: partially working state

ruthra-kumar committed:
refactor: update reconcile flag and setting hard limits for fetching

ruthra-kumar committed:
chore: make allocation editable -- NEED TO REVERT

ruthra-kumar committed:
chore: pause button

ruthra-kumar committed:
refactor: skip setter function in Payment Entry for better performan

ruthra-kumar committed:
refactor: split reconcile function and skip a setter function

1. Split reconcile function into 2
2. While reconciling against payment entry, skip a
set_missing_ref_details setter method

ruthra-kumar committed:
chore: increase payment limit

ruthra-kumar committed:
refactor: replace frappe.db.get_all with frappe.db.get_value

ruthra-kumar committed:
chore: remove unwanted doctypes

ruthra-kumar committed:
refactor: make allocation table readonly

ruthra-kumar committed:
perf: update ref_details only for newly linked invoices

ruthra-kumar committed:
chore: rename skip flag

ruthra-kumar committed:
refactor(UI): receivable_payable field should auto populate

ruthra-kumar committed:
refactor: no control statements in finally block

ruthra-kumar committed:
chore: cleanup section and rename checkbox

ruthra-kumar committed:
chore: update new fieldname in code

ruthra-kumar committed:
chore: update error msg

ruthra-kumar committed:
refactor: start and pause integrated into status

pause checkbox has been removed

ruthra-kumar committed:
refactor: added cancelled status to the log doctype

1. Moved the status section to the bottom in parent doc
2. Using alerts to indicate Job trigger status
